File: app/src/main/java/ru/undframe/needle/tasks/UploadFileTask.java
package ru.undframe.needle.tasks;

import android.os.AsyncTask;
import android.os.Build;

import org.json.JSONObject;

import java.io.File;
import java.io.IOException;
import java.nio.charset.Charset;
import java.util.List;
import java.util.function.Consumer;

import ru.undframe.needle.utils.GlobalProperties;
import ru.undframe.needle.utils.MultipartUtility;

public class UploadFileTask extends AsyncTask<Void, Void, Integer> {

    private File file;
    private Consumer<Integer> action;
    private static File emptyFile;
    static{
        try {
            emptyFile = File.createTempFile("temptile", ".tmp");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }


    public UploadFileTask( File file, Consumer<Integer> action) {
        this.file = file;
        this.action = action;
    }

    @Override
    protected Integer doInBackground(Void... params) {
        try {



                String requestUrl = "http://" + GlobalProperties.SERVER_ADDRESS + "/uploadfile";
                MultipartUtility multipart = new MultipartUtility(requestUrl, Charset.defaultCharset().name());

                if(file!=null)
                    multipart.addFilePart("file", file);


                List<String> response = multipart.finish();

                StringBuilder stringJson = new StringBuilder();

                for (String s : response) {
                    stringJson.append(s);
                }

                JSONObject jsonObject = new JSONObject(stringJson.toString());
                return jsonObject.getInt("status");
        } catch (Exception e) {
            e.printStackTrace();

        }
        return 1;
    }

    @Override
    protected void onPostExecute(Integer result) {
        super.onPostExecute(result);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            action.accept(result);
        }
    }

}

File: app/src/main/java/ru/undframe/needle/tasks/UploadFileTask.kt
package ru.undframe.needle.tasks

import android.os.AsyncTask
import android.os.Build
import android.util.Log
import org.json.JSONObject
import ru.undframe.needle.utils.GlobalProperties.serverAddress
import ru.undframe.needle.utils.MultipartUtility
import java.io.File
import java.io.IOException
import java.nio.charset.Charset
import java.util.function.Consumer

class UploadFileTask(private val accessToken: String, private val file: File) {
    companion object {
        private var emptyFile: File? = null

        init {
            try {
                emptyFile = File.createTempFile("temptile", ".tmp")
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    suspend fun upload(): Int? {
        val requestUrl = "http://${serverAddress}/uploadfile?token=${accessToken}"
        runCatching {
            val multipart = MultipartUtility(requestUrl, Charset.defaultCharset().name())
            multipart.addFilePart("file", file)
            val response = multipart.finish()
            val stringJson = StringBuilder()
            for (s in response) {
                stringJson.append(s)
            }
            val jsonObject = JSONObject(stringJson.toString())
            return jsonObject.getInt("status")
        }.getOrElse { Log.e("Upload file", "from $requestUrl", it) }
        return null
    }
}